cpp/include/algorithms.hpp

```diff
@@ -0,0 +1,65 @@
+ * @tparam WT                        Type of edge weights. Supported value : float or double.   
+ *
+ * @param[in] graph                  cuGRAPH graph descriptor, should contain the connectivity information as a transposed adjacency list (CSR). Edge weights are not used for this algorithm.
+ * @param[in] alpha                  The damping factor alpha represents the probability to follow an outgoing edge, standard value is 0.85.
+                                     Thus, 1.0-alpha is the probability to “teleport” to a random vertex. Alpha should be greater than 0.0 and strictly lower than 1.0.
+ *                                   The initial guess must not be the vector of 0s. Any value other than 1 or 0 is treated as an invalid value.
+ * @param[in] pagerank               Array of size V. Should contain the initial guess if has_guess=true. In this case the initial guess cannot be the vector of 0s. Memory is provided and owned by the caller.
+ * @param[in] personalization_subset_size (optional) The number of vertices for to personalize. Initialized to 0 by default.
+ * @param[in] personalization_subset (optional) Array of size personalization_subset_size containing vertices for running personalized pagerank. Initialized to nullptr by default. Memory is provided and owned by the caller.
+ * @param[in] personalization_values (optional) Array of size personalization_subset_size containing values associated with personalization_subset vertices. Initialized to nullptr by default. Memory is provided and owned by the caller.
+ * @param[in] tolerance              Set the tolerance the approximation, this parameter should be a small magnitude value.
+ *                                   The lower the tolerance the better the approximation. If this value is 0.0f, cuGRAPH will use the default value which is 1.0E-5.
+ *                                   Setting too small a tolerance can lead to non-convergence due to numerical roundoff. Usually values between 0.01 and 0.00001 are acceptable.
+ * @param[in] max_iter               (optional) The maximum number of iterations before an answer is returned. This can be used to limit the execution time and do an early exit before the solver reaches the convergence tolerance.
+ *                                   If this value is lower or equal to 0 cuGRAPH will use the default value, which is 500.
+ * @param[in] has_guess              (optional) This parameter is used to notify cuGRAPH if it should use a user-provided initial guess. False means the user does not have a guess, in this case cuGRAPH will use a uniform vector set to 1/V.
+ *                                   If the value is True, cuGRAPH will read the pagerank parameter and use this as an initial guess.
+ * @param[out] *pagerank             The PageRank : pagerank[i] is the PageRank of vertex i. Memory remains provided and owned by the caller.
+ *
+ */
+template <typename VT, typename ET, typename WT>
```

Comment:
And this might be a bit out-of-scope for this PR,

but we should think more about consistently naming VT, ET, WT.

One concern about this naming is that we're using "value" for "weight" in multiple places. If we consistently use this, then we should use VT (vertex), ET, and VT (value); so we have two VTs.

So, if we decide to consistently use value, we'd better go for vertex_t, edge_t, and value_t.

If we decide to consistently use weight, we'd better avoid using value for weight.

And I don't think pagerank has WT. WT* pagerank or WT* personalization values should have a separate type from edge weights (or edge values).

Reply:
I see two issues here, that we should probably resolve...

1. Our general concept of what a "weighted graph" or "property graph" means, and
1. Consistent naming of these things within the software.

I started a google doc to discuss the first.  I'm inclined to continue that discussion and create a new issue to address any changes we decide to make - unless we can come to a consensus within the next couple of days.

The second issue is something we could grapple with.  I used VT, ET and WT out of Alex's PR that started trying to address this issue.  I tend to prefer less ambiguous names, vertex_t, edge_t and (value_t or weight_t) seem less ambiguous to me.

It might also be beneficial to define some typedefs within the Graph to expose those.  We could consider having the graph type passed into page rank be the templated type and have the parameters from page rank "discover" vertex_t and edge_t from the templated graph object.  That might lead to fewer compiler issues.

I started with what I thought would be the least invasive approach here.

Reply:
Yup, we have multiple issues, and it is impractical to address all those issues in a single PR, just resolve issues you can locally resolve, and for much more involved issues, let's document those and discuss how to resolve those later. I'll just approve this PR, but let me know if you need more reviews.